Final/multinblkserver.cpp

```diff
@@ -0,0 +1,129 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <string.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <netdb.h>
+#include "ThreadPool.h"
+#include "fdlist.h"
+
+const int PORT_NO = 8080;
+const int BACKLOG_SIZE = 2;
+
+// Prototypes
+void select(int sockfd, fd_set *read_fds_ref, struct node *head);
+bool wasEchoed(fd_set *read_fds_ref, struct node *head);
+void echo(int fd);
+int accept(int sockfd);
+
+void error(const char *msg)
+{
+    perror(msg);
+    exit(0);
+}
+
+int main(int argc, char const *argv[])
+{
+	// Step 1: Set up socket
+    int sockfd = socket(AF_INET, SOCK_STREAM, 0); // Create socket file descriptor for server
+    if (sockfd < 0) 
+        error("ERROR opening socket");
+
+	// Step 2: Bind
+	struct sockaddr_in host;
+	host.sin_family = AF_INET; // IPv4 connection
+	host.sin_port = htons(PORT_NO); // bound to port 8080
+	host.sin_addr.s_addr = INADDR_ANY; // connected to localhost
+	const struct sockaddr *host_ptr = (struct sockaddr*) &host;
+	int bound = bind(sockfd, host_ptr, sizeof(host)); // bind sockfd to port
+	if (bound < 0)
+		error("ERROR binding to socket");
+
+	// Step 3: Listen
+	int listening = listen(sockfd, BACKLOG_SIZE); // The socket will actively listen for incoming connections and queue them up
+	if (listening < 0)
+		error("ERROR listener failed");
+
+	fd_set read_fds;
+	struct node *head = createFDList(0); // Create empty fdlist (file descriptor list)
+	while (1) {
+		select(sockfd, &read_fds, head); // select incoming connections and handle them as necessary
+	}
+
+	// Close the connection
+	printf("Closing the connection");
+    close(sockfd);
+	return 0;
+}
+
+//Pre: sockfd is the socket bound to this server and head is a pointer to the beginning of an fdlist
+//Post: sleeps until there is an available connection and then determines if it must be accepted and read from
+void select(int sockfd, fd_set *read_fds_ref, struct node *head) {
+	printf("%s\n", "Attempting to Select . . .");
+	makeFDS(head, read_fds_ref); // make fd_set from fdlist
+	FD_SET(sockfd, read_fds_ref); // (re)add sockfd to allow for new connections
+	int maxsockfd = maxfd(sockfd, head); // determine the maximum value of file descriptor for select
+	int selected = select(maxsockfd, read_fds_ref, NULL, NULL, NULL); // sleep until there has been in a change in one of the file descriptors in the fd_set
+	if (selected >= 0) {
```

Comment:
let's handle the error immediately and explicitly rather than handling it in an else statement:

```
if (select(maxsockfd, read_fds_ref, NULL, NULL, NULL) < 0)
    error("Failed to select");

// Handle connection
```